### ISSUE TITLE
Idea: Improve UX for light clients with covenant on destination address (Liquid only)

### ISSUE THREAD

**Issue by tiero** [sputn1ck/peerswap#121]
DISCLAIMER: I didn't finish reading the spec fully and may have still very little understanding of the system, so bear with me. 🙏 

After a brief discussion with @sputn1ck offline, I come up with a scheme that will allow light clients (ie. browsers/mobile apps) to not require to be always online in order to claim the on-chain funds, delegating to third party services to reveal the `preimage` when opening tx is detected on-chain, without giving up fiduciary custody of funds at any point.

In brief, we can have ad covenant that enforces a specific destination for the on-chain funds committed, in the  in the `claim`  clause.

The benefits is that I can create preimage in the light client, uploading it to a third party service (or multiple of them for redundancy) like cloud functions etc.., going offline.

The third party service will have the duty to monitor the outpoint of the contract the light client is interested to and reveal to the other party so he can claim Lighting funds, on behalf of the light client.

Since the destination output is enforced in the claim transaction, the third party service cannot access funds in any way, can only fail to reveal the preimage triggering the refund path after CSV expires, but if multiple service are run in parallel this can be less of a problem (or maybe incentivized with a small fee)

Let me know if this makes sense, of course only Liquid can enforce destination output with introspection capabilities, so may be a bit of burden to have two different contract code for the two networks.

## example (unblinded output)

```
// let's say the claim tx must always have the destination address in the vout 0
OP_0
OP_INSPECTOUTPUTASSET

//  check that the asset is explicit
OP_1 
OP_EQUALVERIFY

// check the asset ID
<assetID>
OP_EQUALVERIFY

// check value of vout0
OP_0
OP_INSPECTOUTPUTVALUE
//  check that the value is explicit
OP_1 
OP_EQUALVERIFY
<amount_to_be_claimed_64bit>
OP_EQUALVERIFY


// finally check the scriptPubKey destination 
OP_0
OP_INSPECTOUTPUTSCRIPTPUBKEY
// we check the witness version, ie. p2wpkh
OP_0 
OP_EQUALVERIFY
<destination_address>
OP_EQUALVERIFY


// check nonce is empty (unblinded)
OP_0
OP_INSPECTOUTPUTNONCE
OP_0
OP_EQUALVERIFY
```

PS: think may be possible with a blinded outputs, that case we may want to check commitments instead of asset/value revealed values, but need confirmation from blinding wizards.

**Comment by sputn1ck** [comment]
The scheme is definitly pretty cool.

The third party claim service adds some form of trust to the scheme. If the claim service has the preimage, they could collude with the onchain liquidity provice to claim the lightning payment, but wait for the csv to expire.

**Comment by tiero** [comment]
> Meaning they would receive the lightning funds and could keep the onchain funds.

Yes sorry I forgot about this point indeed.

So yes the third party online service not viable, so we may be left to rely on the user's client capabilities of background jobs (service workers on a browser or push notification to wake up mobile phones before csv expires)


I do believe this covenant on the address (dropping CHECKSIG requirement) would still greatly benefit the UX & security of light clients: from a client architecture standpoint this would mean  need the private key to be available by background jobs in order to sign the spending path without user intervention (he "can go away" without having to keep the private key in-memory so to speak)

I can live with the "hey user stay here online until the process is finished" but maybe worth exploring ways to make it less scary UX if we can :)

**Comment by sputn1ck** [comment]
I think ideally we would want to do a taproot keyspend between client and exchange server, in order to minimize cost and maximize privacy. Removing the checksig requirements (that should also work for csv path right?) would be pretty cool. 

FYI, in peerswap we don't use hodl invoices, as we know that we have a payment path to the peer (as we only swap between direct peers) and can block payments over a lightning channel in order for the payment to succeed.

**Comment by nepet** [comment]
> I do believe this covenant on the address (dropping CHECKSIG requirement) would still greatly benefit the UX & security of light clients: from a client architecture standpoint this would mean need the private key to be available by background jobs in order to sign the spending path without user intervention (he "can go away" without having to keep the private key in-memory so to speak)

I see where you are going with this! This way we would need diverging scripts for bitcoin and liquid, right?

**Comment by tiero** [comment]
> I see where you are going with this! This way we would need diverging scripts for bitcoin and liquid, right?

Unfortunately yes :(  I understand that having different script for each chain increases the review surface etc..

Introspection/Covenants only possible on liquid (we may do some experiment with APO/CTV ahead of time, but unlikely introspection opcodes will be added to BTC anyway so divergence could remain)

**Comment by tiero** [comment]
> I think ideally we would want to do a taproot keyspend between client and exchange server

If we want to use introspection opcodes (instead of bare CSFS + CAT tricks) we are forced to use Elements Tapscript indeed.


To not have to diverge by default between Bitcoin and Liquid, we may have two alternative spending paths at the same time  for triggering the `claim_by_invoice` condition if we go for the taproot way:

- Bitcoin-style with  CHECKSIG (as it is now, this can be segwit v0)
- Introspection where we enforce user's destination address (this must be segwit v1, tapscript  spend)

**Comment by sputn1ck** [comment]
> If we want to use introspection opcodes (instead of bare CSFS + CAT tricks) we are forced to use Elements Tapscript indeed.
> 

In my rough understanding of taproot: The public key for the taproot tx could be a 2of2 musig, so that in a cooperative case you would not need to reveal the preimage (and script) onchain, such that only if one party goes offline / does not cooperate you would need to reveal the script. This increases privacy and reduces cost.  Is that right @jonasnick ?

**Comment by tiero** [comment]
Yes this correct in terms of how taproot output would look like (key path spend for the cooperative case, and script path spend for either the claim or the refund). My point is also that on Elements, in order to use the new OPCODES, you must use a segwit v1 address, so yeah that covenant construction needs taproot anyway.
